File: examples/mcp-research-client.js
#!/usr/bin/env node

/**
 * MCP Research Client - Node.js Example
 * 
 * This script demonstrates how to programmatically execute gene research
 * using the Deep Research MCP server API.
 * 
 * Usage:
 *   node mcp-research-client.js
 * 
 * Requirements:
 *   - MCP server running at http://localhost:3000/api/mcp
 *   - Environment variables configured (see .env.local.example)
 */

const https = require('https');
const http = require('http');
const fs = require('fs').promises;
const path = require('path');

// Configuration
const CONFIG = {
  mcpServerUrl: process.env.MCP_SERVER_URL || 'http://localhost:3000/api/mcp',
  accessPassword: process.env.ACCESS_PASSWORD || '',
  timeout: parseInt(process.env.MCP_CLIENT_TIMEOUT || '600000'), // 10 minutes
  outputDir: process.env.OUTPUT_DIR || './research-reports'
};

/**
 * MCP Client for Deep Research
 */
class MCPResearchClient {
  constructor(config) {
    this.config = config;
    this.requestId = 0;
  }

  /**
   * Call MCP tool with given name and arguments
   */
  async callTool(toolName, args) {
    this.requestId++;
    
    const payload = {
      jsonrpc: '2.0',
      id: this.requestId,
      method: 'tools/call',
      params: {
        name: toolName,
        arguments: args
      }
    };

    const headers = {
      'Content-Type': 'application/json'
    };

    if (this.config.accessPassword) {
      headers['Authorization'] = `Bearer ${this.config.accessPassword}`;
    }

    try {
      const result = await this.makeRequest(payload, headers);
      
      if (result.error) {
        throw new Error(`MCP Error: ${result.error.message}`);
      }

      // Parse the response content
      const content = result.result.content[0].text;
      return JSON.parse(content);
    } catch (error) {
      throw new Error(`Tool call failed: ${error.message}`);
    }
  }

  /**
   * Make HTTP request to MCP server
   */
  makeRequest(payload, headers) {
    return new Promise((resolve, reject) => {
      const url = new URL(this.config.mcpServerUrl);
      const isHttps = url.protocol === 'https:';
      const client = isHttps ? https : http;

      const options = {
        hostname: url.hostname,
        port: url.port || (isHttps ? 443 : 80),
        path: url.pathname,
        method: 'POST',
        headers: {
          ...headers,
          'Content-Length': Buffer.byteLength(JSON.stringify(payload))
        },
        timeout: this.config.timeout
      };

      const req = client.request(options, (res) => {
        let data = '';

        res.on('data', (chunk) => {
          data += chunk;
        });

        res.on('end', () => {
          try {
            const parsed = JSON.parse(data);
            resolve(parsed);
          } catch (e) {
            reject(new Error(`Failed to parse response: ${e.message}`));
          }
        });
      });

      req.on('error', (error) => {
        reject(error);
      });

      req.on('timeout', () => {
        req.destroy();
        reject(new Error('Request timeout'));
      });

      req.write(JSON.stringify(payload));
      req.end();
    });
  }

  /**
   * Conduct complete gene research (recommended)
   */
  async conductGeneResearch(options) {
    const {
      geneSymbol,
      organism,
      researchFocus = [],
      specificAspects = [],
      diseaseContext,
      experimentalApproach,
      userPrompt,
      language = 'en-US',
      maxResult = 10,
      enableCitationImage = true,
      enableReferences = true
    } = options;

    console.log(`\n${'='.repeat(60)}`);
    console.log(`🔬 Starting Gene Research`);
    console.log(`${'='.repeat(60)}`);
    console.log(`Gene: ${geneSymbol}`);
    console.log(`Organism: ${organism}`);
    if (researchFocus.length > 0) {
      console.log(`Focus: ${researchFocus.join(', ')}`);
    }
    console.log(`Max Results: ${maxResult}`);
    console.log('');

    const startTime = Date.now();

    try {
      const result = await this.callTool('gene-research', {
        geneSymbol,
        organism,
        researchFocus,
        specificAspects,
        diseaseContext,
        experimentalApproach,
        userPrompt,
        language,
        maxResult,
        enableCitationImage,
        enableReferences
      });

      const duration = ((Date.now() - startTime) / 1000).toFixed(2);

      console.log('✅ Research Completed!\n');
      console.log('📊 Results:');
      console.log(`   - Quality Score: ${(result.qualityMetrics.overallQuality * 100).toFixed(1)}%`);
      console.log(`   - Data Completeness: ${(result.metadata.completeness * 100).toFixed(1)}%`);
      console.log(`   - Confidence: ${(result.metadata.confidence * 100).toFixed(1)}%`);
      console.log(`   - Research Time: ${duration}s`);
      console.log(`   - Data Sources: ${result.metadata.dataSources.join(', ')}`);
      console.log('');

      return result;
    } catch (error) {
      console.error(`❌ Research failed: ${error.message}`);
      throw error;
    }
  }

  /**
   * Step-by-step research workflow
   */
  async stepByStepResearch(query, language = 'en-US', maxResult = 10) {
    console.log(`\n${'='.repeat(60)}`);
    console.log(`📋 Step-by-Step Research Workflow`);
    console.log(`${'='.repeat(60)}\n`);

    try {
      // Step 1: Generate research plan
      console.log('📝 Step 1: Generating research plan...');
      const planResult = await this.callTool('write-research-plan', {
        query,
        language
      });
      console.log(`✅ Plan generated (${planResult.reportPlan.length} chars)\n`);

      // Step 2: Generate search queries
      console.log('🔍 Step 2: Generating search tasks...');
      const searchTasks = await this.callTool('generate-SERP-query', {
        plan: planResult.reportPlan,
        language
      });
      console.log(`✅ Generated ${searchTasks.length} search tasks\n`);

      // Step 3: Execute searches
      console.log(`📚 Step 3: Executing ${searchTasks.length} searches...`);
      const completedTasks = await this.callTool('search-task', {
        tasks: searchTasks,
        language,
        maxResult,
        enableReferences: true
      });
      
      const totalSources = completedTasks.reduce(
        (sum, task) => sum + (task.sources?.length || 0),
        0
      );
      console.log(`✅ Collected ${totalSources} sources\n`);

      // Step 4: Generate final report
      console.log('📄 Step 4: Generating final report...');
      const report = await this.callTool('write-final-report', {
        plan: planResult.reportPlan,
        tasks: completedTasks,
        language,
        maxResult,
        enableCitationImage: true,
        enableReferences: true
      });
      console.log('✅ Report generated!\n');

      return {
        plan: planResult.reportPlan,
        tasks: completedTasks,
        report
      };
    } catch (error) {
      console.error(`❌ Workflow failed: ${error.message}`);
      throw error;
    }
  }

  /**
   * Save research report to file
   */
  async saveReport(geneSymbol, organism, reportContent, outputDir) {
    try {
      // Ensure output directory exists
      await fs.mkdir(outputDir, { recursive: true });

      // Generate filename
      const timestamp = new Date().toISOString().replace(/[:.]/g, '-').split('T')[0];
      const safeGene = geneSymbol.replace(/[^a-zA-Z0-9]/g, '_');
      const safeOrg = organism.replace(/[^a-zA-Z0-9]/g, '_');
      const filename = `${safeGene}_${safeOrg}_${timestamp}.md`;
      const filepath = path.join(outputDir, filename);

      // Write report
      await fs.writeFile(filepath, reportContent, 'utf-8');

      console.log(`💾 Report saved: ${filepath}`);
      return filepath;
    } catch (error) {
      console.error(`Failed to save report: ${error.message}`);
      throw error;
    }
  }

  /**
   * Save complete research data (JSON)
   */
  async saveResearchData(geneSymbol, organism, data, outputDir) {
    try {
      await fs.mkdir(outputDir, { recursive: true });

      const timestamp = new Date().toISOString().replace(/[:.]/g, '-').split('T')[0];
      const safeGene = geneSymbol.replace(/[^a-zA-Z0-9]/g, '_');
      const safeOrg = organism.replace(/[^a-zA-Z0-9]/g, '_');
      const filename = `${safeGene}_${safeOrg}_${timestamp}_data.json`;
      const filepath = path.join(outputDir, filename);

      await fs.writeFile(filepath, JSON.stringify(data, null, 2), 'utf-8');

      console.log(`💾 Research data saved: ${filepath}`);
      return filepath;
    } catch (error) {
      console.error(`Failed to save research data: ${error.message}`);
      throw error;
    }
  }
}

/**
 * Main execution function
 */
async function main() {
  const client = new MCPResearchClient(CONFIG);

  try {
    // Example 1: Complete gene research (recommended)
    const researchResult = await client.conductGeneResearch({
      geneSymbol: 'talB',
      organism: 'Escherichia coli',
      researchFocus: [
        'molecular_function',
        'metabolic_pathways',
        'protein_structure',
        'regulation'
      ],
      specificAspects: [
        'transaldolase mechanism',
        'pentose phosphate pathway',
        'cofactor requirements'
      ],
      language: 'en-US',
      maxResult: 10,
      enableReferences: true
    });

    // Save report - prioritize finalReport (the complete markdown content)
    const reportContent = researchResult.finalReport || 
                         researchResult.report?.content || 
                         (researchResult.report?.sections ? 
                           researchResult.report.sections.map(s => s.content).join('\n\n') : 
                           '');
    
    if (reportContent) {
      await client.saveReport(
        'talB',
        'Escherichia coli',
        reportContent,
        CONFIG.outputDir
      );
    } else {
      console.log('⚠️  No report content available to save');
    }

    // Save complete data
    await client.saveResearchData(
      'talB',
      'Escherichia coli',
      researchResult,
      CONFIG.outputDir
    );

    console.log('\n✨ Research completed successfully!\n');
    
    // Display summary
    console.log('📋 Summary:');
    console.log(`   - Report available: ${reportContent ? 'Yes' : 'No'}`);
    console.log(`   - Report length: ${reportContent ? reportContent.length + ' chars' : 'N/A'}`);
    console.log(`   - Visualizations: ${researchResult.visualizations?.length || researchResult.geneResearch?.visualizations?.length || 0}`);
    console.log(`   - Sources: ${researchResult.sources?.length || 0}`);
    console.log(`   - Quality metrics available: Yes`);
    console.log('');

  } catch (error) {
    console.error(`\n❌ Error: ${error.message}\n`);
    process.exit(1);
  }
}

/**
 * Batch research example
 */
async function batchResearch() {
  const client = new MCPResearchClient(CONFIG);

  const genes = [
    {
      geneSymbol: 'talB',
      organism: 'Escherichia coli',
      researchFocus: ['molecular_function', 'metabolic_pathways']
    },
    {
      geneSymbol: 'lysC',
      organism: 'Escherichia coli',
      researchFocus: ['enzyme_kinetics', 'regulation']
    },
    {
      geneSymbol: 'thrB',
      organism: 'Escherichia coli',
      researchFocus: ['protein_structure', 'allosteric_regulation']
    }
  ];

  console.log(`\n${'='.repeat(60)}`);
  console.log(`🔬 Batch Gene Research (${genes.length} genes)`);
  console.log(`${'='.repeat(60)}\n`);

  const results = [];

  for (let i = 0; i < genes.length; i++) {
    const gene = genes[i];
    console.log(`\n[${ i + 1}/${genes.length}] Processing ${gene.geneSymbol}...`);

    try {
      const result = await client.conductGeneResearch({
        ...gene,
        language: 'en-US',
        maxResult: 10,
        enableReferences: true
      });

      const reportContent = result.report?.content || result.finalReport || '';
      
      if (reportContent) {
        await client.saveReport(
          gene.geneSymbol,
          gene.organism,
          reportContent,
          CONFIG.outputDir
        );
      }

      results.push({ gene: gene.geneSymbol, success: true, result });
      
      // Add delay between requests to avoid overwhelming the server
      if (i < genes.length - 1) {
        console.log('⏳ Waiting 5s before next request...');
        await new Promise(resolve => setTimeout(resolve, 5000));
      }
    } catch (error) {
      console.error(`Failed: ${error.message}`);
      results.push({ gene: gene.geneSymbol, success: false, error: error.message });
    }
  }

  // Summary
  console.log(`\n${'='.repeat(60)}`);
  console.log('📊 Batch Research Summary');
  console.log(`${'='.repeat(60)}\n`);
  
  const successful = results.filter(r => r.success).length;
  const failed = results.filter(r => !r.success).length;
  
  console.log(`✅ Successful: ${successful}/${genes.length}`);
  console.log(`❌ Failed: ${failed}/${genes.length}`);
  console.log('');

  results.forEach(r => {
    const status = r.success ? '✅' : '❌';
    console.log(`${status} ${r.gene}: ${r.success ? 'Success' : r.error}`);
  });
  console.log('');
}

// Export for use as module
module.exports = { MCPResearchClient, CONFIG };

// Run main function if executed directly
if (require.main === module) {
  // Check command line arguments
  const args = process.argv.slice(2);
  
  if (args.includes('--batch')) {
    batchResearch().catch(error => {
      console.error(error);
      process.exit(1);
    });
  } else if (args.includes('--help') || args.includes('-h')) {
    console.log(`
Usage: node mcp-research-client.js [options]

Options:
  --batch          Run batch research on multiple genes
  --help, -h       Show this help message

Environment Variables:
  MCP_SERVER_URL        MCP server URL (default: http://localhost:3000/api/mcp)
  ACCESS_PASSWORD       Access password for authentication
  MCP_CLIENT_TIMEOUT    Request timeout in ms (default: 600000)
  OUTPUT_DIR            Output directory for reports (default: ./research-reports)

Examples:
  # Single gene research
  node mcp-research-client.js

  # Batch research
  node mcp-research-client.js --batch

  # With custom configuration
  MCP_SERVER_URL=http://example.com/api/mcp ACCESS_PASSWORD=secret node mcp-research-client.js
`);
  } else {
    main().catch(error => {
      console.error(error);
      process.exit(1);
    });
  }
}

File: examples/mcp-research-client.py
#!/usr/bin/env python3
"""
MCP Research Client - Python Example

This script demonstrates how to programmatically execute gene research
using the Deep Research MCP server API.

Usage:
    python mcp-research-client.py
    python mcp-research-client.py --batch

Requirements:
    pip install httpx
"""

import asyncio
import json
import os
import sys
from pathlib import Path
from datetime import datetime
from typing import Dict, List, Optional, Any

try:
    import httpx
except ImportError:
    print("Error: httpx is required. Install it with: pip install httpx")
    sys.exit(1)


# Configuration
class Config:
    MCP_SERVER_URL = os.getenv('MCP_SERVER_URL', 'http://localhost:3000/api/mcp')
    ACCESS_PASSWORD = os.getenv('ACCESS_PASSWORD', '')
    MCP_CLIENT_TIMEOUT = int(os.getenv('MCP_CLIENT_TIMEOUT', '600'))
    OUTPUT_DIR = os.getenv('OUTPUT_DIR', './research-reports')


class MCPResearchClient:
    """Client for Deep Research MCP server"""
    
    def __init__(self, config: Optional[Config] = None):
        self.config = config or Config
        self.request_id = 0
        self.headers = {
            'Content-Type': 'application/json'
        }
        
        if self.config.ACCESS_PASSWORD:
            self.headers['Authorization'] = f'Bearer {self.config.ACCESS_PASSWORD}'
    
    async def call_tool(self, tool_name: str, arguments: Dict[str, Any]) -> Any:
        """Call MCP tool with given name and arguments"""
        self.request_id += 1
        
        payload = {
            'jsonrpc': '2.0',
            'id': self.request_id,
            'method': 'tools/call',
            'params': {
                'name': tool_name,
                'arguments': arguments
            }
        }
        
        async with httpx.AsyncClient(timeout=self.config.MCP_CLIENT_TIMEOUT) as client:
            try:
                response = await client.post(
                    self.config.MCP_SERVER_URL,
                    json=payload,
                    headers=self.headers
                )
                
                result = response.json()
                
                if 'error' in result:
                    raise Exception(f"MCP Error: {result['error']['message']}")
                
                # Parse the response content
                content = result['result']['content'][0]['text']
                return json.loads(content)
                
            except httpx.TimeoutException:
                raise Exception('Request timeout')
            except Exception as e:
                raise Exception(f'Tool call failed: {str(e)}')
    
    async def conduct_gene_research(
        self,
        gene_symbol: str,
        organism: str,
        research_focus: Optional[List[str]] = None,
        specific_aspects: Optional[List[str]] = None,
        disease_context: Optional[str] = None,
        experimental_approach: Optional[str] = None,
        user_prompt: Optional[str] = None,
        language: str = 'en-US',
        max_result: int = 10,
        enable_citation_image: bool = True,
        enable_references: bool = True
    ) -> Dict[str, Any]:
        """Conduct complete gene research (recommended)"""
        
        print(f"\n{'='*60}")
        print(f"🔬 Starting Gene Research")
        print(f"{'='*60}")
        print(f"Gene: {gene_symbol}")
        print(f"Organism: {organism}")
        
        if research_focus:
            print(f"Focus: {', '.join(research_focus)}")
        
        print(f"Max Results: {max_result}")
        print()
        
        start_time = datetime.now()
        
        try:
            result = await self.call_tool('deep-gene-research', {
                'geneSymbol': gene_symbol,
                'organism': organism,
                'researchFocus': research_focus or [],
                'specificAspects': specific_aspects or [],
                'diseaseContext': disease_context,
                'experimentalApproach': experimental_approach,
                'userPrompt': user_prompt,
                'language': language,
                'maxResult': max_result,
                'enableCitationImage': enable_citation_image,
                'enableReferences': enable_references
            })
            
            duration = (datetime.now() - start_time).total_seconds()
            
            print('✅ Research Completed!\n')
            print('📊 Results:')
            print(f"   - Quality Score: {result['qualityMetrics']['overallQuality']*100:.1f}%")
            print(f"   - Data Completeness: {result['metadata']['completeness']*100:.1f}%")
            print(f"   - Confidence: {result['metadata']['confidence']*100:.1f}%")
            print(f"   - Research Time: {duration:.2f}s")
            print(f"   - Data Sources: {', '.join(result['metadata']['dataSources'])}")
            print()
            
            return result
            
        except Exception as e:
            print(f"❌ Research failed: {str(e)}")
            raise
    
    async def step_by_step_research(
        self,
        query: str,
        language: str = 'en-US',
        max_result: int = 10
    ) -> Dict[str, Any]:
        """Step-by-step research workflow"""
        
        print(f"\n{'='*60}")
        print(f"📋 Step-by-Step Research Workflow")
        print(f"{'='*60}\n")
        
        try:
            # Step 1: Generate research plan
            print('📝 Step 1: Generating research plan...')
            plan_result = await self.call_tool('write-research-plan', {
                'query': query,
                'language': language
            })
            print(f"✅ Plan generated ({len(plan_result['reportPlan'])} chars)\n")
            
            # Step 2: Generate search queries
            print('🔍 Step 2: Generating search tasks...')
            search_tasks = await self.call_tool('generate-SERP-query', {
                'plan': plan_result['reportPlan'],
                'language': language
            })
            print(f"✅ Generated {len(search_tasks)} search tasks\n")
            
            # Step 3: Execute searches
            print(f"📚 Step 3: Executing {len(search_tasks)} searches...")
            completed_tasks = await self.call_tool('search-task', {
                'tasks': search_tasks,
                'language': language,
                'maxResult': max_result,
                'enableReferences': True
            })
            
            total_sources = sum(len(task.get('sources', [])) for task in completed_tasks)
            print(f"✅ Collected {total_sources} sources\n")
            
            # Step 4: Generate final report
            print('📄 Step 4: Generating final report...')
            report = await self.call_tool('write-final-report', {
                'plan': plan_result['reportPlan'],
                'tasks': completed_tasks,
                'language': language,
                'maxResult': max_result,
                'enableCitationImage': True,
                'enableReferences': True
            })
            print('✅ Report generated!\n')
            
            return {
                'plan': plan_result['reportPlan'],
                'tasks': completed_tasks,
                'report': report
            }
            
        except Exception as e:
            print(f"❌ Workflow failed: {str(e)}")
            raise
    
    async def save_report(
        self,
        gene_symbol: str,
        organism: str,
        report_content: str,
        output_dir: Optional[str] = None
    ) -> str:
        """Save research report to file"""
        
        output_dir = output_dir or self.config.OUTPUT_DIR
        Path(output_dir).mkdir(parents=True, exist_ok=True)
        
        # Generate filename
        timestamp = datetime.now().strftime('%Y-%m-%d')
        safe_gene = gene_symbol.replace('/', '_').replace(' ', '_')
        safe_org = organism.replace('/', '_').replace(' ', '_')
        filename = f"{safe_gene}_{safe_org}_{timestamp}.md"
        filepath = Path(output_dir) / filename
        
        # Write report
        filepath.write_text(report_content, encoding='utf-8')
        
        print(f"💾 Report saved: {filepath}")
        return str(filepath)
    
    async def save_research_data(
        self,
        gene_symbol: str,
        organism: str,
        data: Dict[str, Any],
        output_dir: Optional[str] = None
    ) -> str:
        """Save complete research data (JSON)"""
        
        output_dir = output_dir or self.config.OUTPUT_DIR
        Path(output_dir).mkdir(parents=True, exist_ok=True)
        
        # Generate filename
        timestamp = datetime.now().strftime('%Y-%m-%d')
        safe_gene = gene_symbol.replace('/', '_').replace(' ', '_')
        safe_org = organism.replace('/', '_').replace(' ', '_')
        filename = f"{safe_gene}_{safe_org}_{timestamp}_data.json"
        filepath = Path(output_dir) / filename
        
        # Write data
        filepath.write_text(json.dumps(data, indent=2), encoding='utf-8')
        
        print(f"💾 Research data saved: {filepath}")
        return str(filepath)


async def main():
    """Main execution function"""
    
    client = MCPResearchClient()
    
    try:
        # Example: Complete gene research (recommended)
        result = await client.conduct_gene_research(
            gene_symbol='talB',
            organism='Escherichia coli',
            research_focus=[
                'molecular_function',
                'metabolic_pathways',
                'protein_structure',
                'regulation'
            ],
            specific_aspects=[
                'transaldolase mechanism',
                'pentose phosphate pathway',
                'cofactor requirements'
            ],
            language='en-US',
            max_result=10,
            enable_references=True
        )
        
        # Save report
        report_content = result.get('report', {}).get('content') or result.get('finalReport') or ''
        
        if report_content:
            await client.save_report(
                'talB',
                'Escherichia coli',
                report_content
            )
        else:
            print('⚠️  No report content available to save')
        
        # Save complete data
        await client.save_research_data(
            'talB',
            'Escherichia coli',
            result
        )
        
        print('\n✨ Research completed successfully!\n')
        
        # Display summary
        print('📋 Summary:')
        print(f"   - Report available: {'Yes' if report_content else 'No'}")
        print(f"   - Report length: {len(report_content) if report_content else 'N/A'} chars")
        print(f"   - Visualizations: {len(result.get('visualizations', result.get('geneResearch', {}).get('visualizations', [])))}")
        print(f"   - Sources: {len(result.get('sources', []))}")
        print(f"   - Quality metrics available: Yes")
        print()
        
    except Exception as e:
        print(f"\n❌ Error: {str(e)}\n")
        sys.exit(1)


async def batch_research():
    """Batch research example"""
    
    client = MCPResearchClient()
    
    genes = [
        {
            'geneSymbol': 'talB',
            'organism': 'Escherichia coli',
            'researchFocus': ['molecular_function', 'metabolic_pathways']
        },
        {
            'geneSymbol': 'lysC',
            'organism': 'Escherichia coli',
            'researchFocus': ['enzyme_kinetics', 'regulation']
        },
        {
            'geneSymbol': 'thrB',
            'organism': 'Escherichia coli',
            'researchFocus': ['protein_structure', 'allosteric_regulation']
        }
    ]
    
    print(f"\n{'='*60}")
    print(f"🔬 Batch Gene Research ({len(genes)} genes)")
    print(f"{'='*60}\n")
    
    results = []
    
    for i, gene_config in enumerate(genes):
        print(f"\n[{i+1}/{len(genes)}] Processing {gene_config['geneSymbol']}...")
        
        try:
            result = await client.conduct_gene_research(
                **gene_config,
                language='en-US',
                max_result=10,
                enable_references=True
            )
            
            report_content = result.get('report', {}).get('content') or result.get('finalReport') or ''
            
            if report_content:
                await client.save_report(
                    gene_config['geneSymbol'],
                    gene_config['organism'],
                    report_content
                )
            
            results.append({
                'gene': gene_config['geneSymbol'],
                'success': True,
                'result': result
            })
            
            # Add delay between requests
            if i < len(genes) - 1:
                print('⏳ Waiting 5s before next request...')
                await asyncio.sleep(5)
                
        except Exception as e:
            print(f"Failed: {str(e)}")
            results.append({
                'gene': gene_config['geneSymbol'],
                'success': False,
                'error': str(e)
            })
    
    # Summary
    print(f"\n{'='*60}")
    print('📊 Batch Research Summary')
    print(f"{'='*60}\n")
    
    successful = sum(1 for r in results if r['success'])
    failed = len(results) - successful
    
    print(f"✅ Successful: {successful}/{len(genes)}")
    print(f"❌ Failed: {failed}/{len(genes)}")
    print()
    
    for r in results:
        status = '✅' if r['success'] else '❌'
        msg = 'Success' if r['success'] else r.get('error', 'Unknown error')
        print(f"{status} {r['gene']}: {msg}")
    print()


if __name__ == '__main__':
    # Check command line arguments
    if '--batch' in sys.argv:
        asyncio.run(batch_research())
    elif '--help' in sys.argv or '-h' in sys.argv:
        print("""
Usage: python mcp-research-client.py [options]

Options:
  --batch          Run batch research on multiple genes
  --help, -h       Show this help message

Environment Variables:
  MCP_SERVER_URL        MCP server URL (default: http://localhost:3000/api/mcp)
  ACCESS_PASSWORD       Access password for authentication
  MCP_CLIENT_TIMEOUT    Request timeout in seconds (default: 600)
  OUTPUT_DIR            Output directory for reports (default: ./research-reports)

Examples:
  # Single gene research
  python mcp-research-client.py

  # Batch research
  python mcp-research-client.py --batch

  # With custom configuration
  MCP_SERVER_URL=http://example.com/api/mcp ACCESS_PASSWORD=secret python mcp-research-client.py
""")
    else:
        asyncio.run(main())
